free5gc-operator/controllers/smfdeployment_controller.go

```diff
@@ -0,0 +1,648 @@
+		memoryRequest = "512Mi"
+		memoryLimit = "512Mi"
+	}
+
+	resources := apiv1.ResourceRequirements{}
+	resources.Limits = make(apiv1.ResourceList)
+	resources.Limits[apiv1.ResourceCPU] = resource.MustParse(cpuLimit)
+	resources.Limits[apiv1.ResourceMemory] = resource.MustParse(memoryLimit)
+
+	resources.Requests = make(apiv1.ResourceList)
+	resources.Requests[apiv1.ResourceCPU] = resource.MustParse(cpuRequest)
+	resources.Requests[apiv1.ResourceMemory] = resource.MustParse(memoryRequest)
+
+	return replicas, &resources, nil
+}
+
+func constructSMFNadName(templateName string, suffix string) string {
+	return templateName + "-" + suffix
+	// return "n4"
+}
+
+// getSMFNad returns NAD label string composed based on the Nx interfaces configuration provided in SMFDeploymentSpec
+func getSMFNad(templateName string, spec *workloadv1alpha1.SMFDeploymentSpec) string {
+	var ret string
+
+	// n6CfgSlice := getIntConfigSlice(spec.Interfaces, "N6")
+	// n3CfgSlice := getIntConfigSlice(spec.Interfaces, "N3")
+	n4CfgSlice := getIntConfigSlice(spec.Interfaces, "n4")
+	// n9CfgSlice := getIntConfigSlice(spec.Interfaces, "N9")
+
+	ret = `[`
+	intfMap := map[string][]workloadv1alpha1.InterfaceConfig{
+		// "n3": n3CfgSlice,
+		"n4": n4CfgSlice,
+		// "n6": n6CfgSlice,
+		// "n9": n9CfgSlice,
+	}
+	// Need to sort inftMap by key otherwise unitTests might fail as order of intefaces in intfMap is not guaranteed
+	inftMapKeys := make([]string, 0, len(intfMap))
+	for interfaceName := range intfMap {
+		inftMapKeys = append(inftMapKeys, interfaceName)
+	}
+	sort.Strings(inftMapKeys)
+
+	noComma := true
+	for _, key := range inftMapKeys {
+		for _, intf := range intfMap[key] {
+			newNad := fmt.Sprintf(`
+        {"name": "%s",
+         "interface": "%s",
+         "ips": ["%s"],
+         "gateways": ["%s"]
+        }`, constructSMFNadName(templateName, key), intf.Name, intf.IPv4.Address, *intf.IPv4.Gateway)
+			if noComma {
+				ret = ret + newNad
+				noComma = false
+			} else {
+				ret = ret + "," + newNad
+			}
+		}
+	}
+	ret = ret + `
+    ]`
+	return ret
+
+	// return `[
+	//     {"name": "free5gc-smf-1-n4",
+	//      "interface": "n4",
+	//     },
+	// ]`
+
+}
+
+// checkNADExists gets deployment object and checks "k8s.v1.cni.cncf.io/networks" NADs.
+// returns True if all requred NADs are present
+// returns False if any NAD doesn't exists in deployment namespace
+func (r *SMFDeploymentReconciler) checkSMFNADexist(log logr.Logger, ctx context.Context, deployment *appsv1.Deployment) bool {
+	smfAnnotations := []SMFAnnotation{}
+	annotationsString, ok := deployment.Spec.Template.GetAnnotations()["k8s.v1.cni.cncf.io/networks"]
+	if !ok {
+		log.Info("Annotations k8s.v1.cni.cncf.io/networks not found", "SMFDeployment.namespace", deployment.Namespace)
+		return false
+	}
+	if err := json.Unmarshal([]byte(annotationsString), &smfAnnotations); err != nil {
+		log.Info("Failed to parse SMFDeployment annotations", "SMFDeployment.namespace", deployment.Namespace)
```

Comment:
So, this is an error, not a false return. The method should return (bool, error) and in this case return the error.

We shouldn't need to log here either - even if there is an error. I would expect the caller to handle the error (which may include logging or something else).